tests/common/type_list.h

```diff
@@ -33,37 +33,131 @@ struct user_struct {
+    return *this;
+  }
+
+  friend bool operator==(const no_cnstr &lhs, const no_cnstr &rhs) {
+    return ((lhs.a == rhs.a) && (lhs.b == rhs.b) && (lhs.c == rhs.c));
+  }
+};
+
+// A user-defined class with several scalar member variables, a user-defined
+//  default constructor, and some member functions that modify the member
+//  variables.
+struct def_cnstr {
+  float a;
+  int b;
+  char c;
+
+ public:
+  constexpr def_cnstr() : a(3.0), b(2), c('c') {}
+
+  constexpr void assign(int val) {
+    a = val * 3.0;
+    b = val * 2;
+    c = val;
+  }
+
+  constexpr auto &operator=(const int &v) {
+    assign(v);
+    return *this;
+  }
+
+  inline friend bool operator==(const def_cnstr &lhs, const def_cnstr &rhs) {
+    return ((lhs.a == rhs.a) && (lhs.b == rhs.b) && (lhs.c == rhs.c));
+  }
+};
+
+// A user-defined class with several scalar member variables, a deleted default
+// constructor, and a user-defined (non-default) constructor.
+class no_def_cnstr {
+  float a;
+  int b;
+  char c;
+
+ public:
+  no_def_cnstr() = delete;
+
+  constexpr no_def_cnstr(int val) : a(val * 3.0), b(val * 2), c(val) {}
+
+  friend bool operator==(const no_def_cnstr &lhs, const no_def_cnstr &rhs) {
+    return ((lhs.a == rhs.a) && (lhs.b == rhs.b) && (lhs.c == rhs.c));
+  }
+};
+// Returns instance of type T
+template <typename T>
+inline constexpr auto get_init_value_helper(int x) {
+  return x;
+}
+
+// Returns instance of type bool
+template <>
+inline constexpr auto get_init_value_helper<bool>(int x) {
+  return (x % 2 != 0);
+}
+
+// Returns instance of user defined struct with no constructor
+template <>
+inline constexpr auto get_init_value_helper<no_cnstr>(int x) {
+  no_cnstr instance{};
+  instance = x;
+  return instance;
+}
+
+// Returns instance of user defined struct default constructor
+template <>
+inline constexpr auto get_init_value_helper<def_cnstr>(int x) {
+  def_cnstr instance;
+  instance = x;
+  return instance;
+}
+
+}  // namespace user_def_types
+
 namespace get_cts_types {
-static const auto vector_types = named_type_pack<
-    bool, char, signed char, unsigned char, short, unsigned short, int,
-    unsigned int, long, unsigned long, long long, unsigned long long, float,
-    sycl::cl_float, sycl::byte, sycl::cl_bool, sycl::cl_char,
-    sycl::cl_uchar, sycl::cl_short, sycl::cl_ushort,
-    sycl::cl_int, sycl::cl_uint, sycl::cl_long, sycl::cl_ulong>{
-    "bool",
-    "char",
-    "signed char",
-    "unsigned char",
-    "short",
-    "unsigned short",
-    "int",
-    "unsigned int",
-    "long",
-    "unsigned long",
-    "long long",
-    "unsigned long long",
-    "float",
-    "sycl::cl_float",
-    "sycl::byte",
-    "sycl::cl_bool",
-    "sycl::cl_char",
-    "sycl::cl_uchar",
-    "sycl::cl_short",
-    "sycl::cl_ushort",
-    "sycl::cl_int",
-    "sycl::cl_uint",
-    "sycl::cl_long",
-    "sycl::cl_ulong"};
-} // namespace get_cts_type
-
-#endif // __SYCLCTS_TESTS_COMMON_TYPE_LIST_H
+static const auto vector_types =
```

Comment:
Shouldn't these global variables be `inline`?

Reply:
I think it will be better if we use a function that returns `vector_types` variable.
I applied this change in 963504b390bbb87f3c01063d8f1c389c30597a59